Test adc support temperature sensor stm32 mcu

TDhaouST opened the pull request:
The goal is to test the junction temperature sensor and the
vref int property in the ADC device-tree node of each stm32 where it is
available.

Check all the ADC of the stm32 mcus where the temp and VrefInt monitoring
is available (based on the Ref Man).
Check that has-temp-channel; and has-vref-channel; in the corresponding
ADC node of the DTS of each stm32 mcu is correctly set.
Verify the VTEMP/ VREFINT activation in the in adc_stm32.c for example).
It is necessary to add and activate the node die-temp in the boards dts files.
Execute the testcase running the samples/sensor/stm32_temp_sensor on
each available target board.

TDhaouST committed:
dts: arm: st: Add adc support of the temperature sensor on stm32

Add the support of the temperature sensor in the ADC device-tree node of
each stm32 where it is available.
- Check all the ADC of the stm32 mcus where the temp and VrefInt monitoring
is available (based on the Ref Man).
- Check that has-temp-channel; and has-vref-channel; in the corresponding
ADC node of the DTS of each stm32 mcu is correctly set.
- Verify the VTEMP/ VREFINT activation in the in adc_stm32.c for example).
-Add the die-temp node (based on the ref man/ datasheet).

Signed-off-by: TLIG Dhaou <dhaou.tlig-ext@st.com>

TDhaouST committed:
boards: arm: Enable the die-temp capabilities of ADC

-Enable the die-temp capabilities of ADC in each dts file of stm32
 boards where it is available.
-Execute the testcase running the samples/sensor/stm32_temp_sensor on each
 available target board.

Signed-off-by: TLIG Dhaou <dhaou.tlig-ext@st.com>

ABOSTM commented on samples/sensor/stm32_temp_sensor/sample.yaml:
```diff
@@ -5,5 +5,10 @@ tests:
+    harness_config:
+      type: one_line
+      regex:
+        - r"Current temperature ^([0-9]{1,2}){1}(\.[0-9]{1,2}) °C"
```
Modifications on sample part should be part of a separate commit.

I think that Regex could be simplified:
r"Current temperature [0-9]{1,2}\.[0-9]{1,2} °C"

erwango replied:
+1 for the separate commit.
Regarding regex, let's limit expected t° range to 10>60.
Question on the "floating" part: Is it possible to get 2 digit ?

TDhaouST replied:
According the example, the float part is composed of a single digit.

ABOSTM reviewed (changes requested):
Missing doc update for nucleo_g0b1 and stm32f030_demo

TDhaouST committed:
Samples: sensor: stm32_temp_sensor: Edit the yaml file of the sample

-Edit the yaml file of the sample in order to execute tests.

Signed-off-by: TLIG Dhaou <dhaou.tlig-ext@st.com>

ABOSTM commented on samples/sensor/stm32_temp_sensor/sample.yaml:
```diff
@@ -5,5 +5,10 @@ tests:
+    harness_config:
+      type: one_line
+      regex:
+        - "Current temperature: [1-5][0-9].[0-9]{1} °C"
```
```suggestion
        - "Current temperature: [1-5][0-9].[0-9] °C"
```

erwango replied:
@TDhaouST Can you take this request into account ?

FRASTM commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -86,6 +86,16 @@
 		};
 
+		die_temp: dietemp {
+			label = "DIE_TEMP";
```
to be checked : stm32g4 has calibration

FRASTM commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -736,6 +736,17 @@
+		};
+
+		die_temp: dietemp {
+			label = "DIE_TEMP";
```
to be checked : stm32h7 has calibration (ex of stm32743)

TDhaouST committed:
drivers: adc: stm32 init the Regular group for stm32u5

The stm32u5 mcu requires the REG group of the ADC
to be initialized before starting the conversion.

Signed-off-by: TLIG Dhaou <dhaou.tlig-ext@st.com>

TDhaouST committed:
drivers: adc: stm32 init the Regular group for stm32wl

The stm32wl mcu requires the REG group of the ADC
to be initialized before starting the conversion.

Signed-off-by: TLIG Dhaou <dhaou.tlig-ext@st.com>

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -88,11 +88,37 @@ static int stm32_temp_channel_get(const struct device *dev, enum sensor_channel
 
 #if HAS_CALIBRATION
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	/*
+	 * TS_CAL are 16bit values (cf Data Sheet), they are divided to match the 12bit
+	 * resolution of the data_raw.
+	 * The 12bit resolution is set by the asp.resolution in stm32_temp_init()
+	 */
+	temp = ((float)data->raw * cfg->tsv_mv) / cfg->cal_vrefanalog;
+	temp -= (*cfg->cal1_addr >> 4);
+	temp *= cfg->cal2_temp - cfg->cal1_temp;
+	temp /= (*cfg->cal2_addr >> 4) - (*cfg->cal1_addr >> 4);
+	temp += cfg->cal_offset;
+
+#elif defined(CONFIG_SOC_SERIES_STM32U5X)
+	/*
+	 * TS_CAL are 14bit values (cf Data Sheet), they are divided to match the 12bit
+	 * resolution of the data_raw.
+	 * The 12bit resolution is set by the asp.resolution in stm32_temp_init()
+	 */
+	temp = ((float)data->raw * cfg->tsv_mv) / cfg->cal_vrefanalog;
+	temp -= (*cfg->cal1_addr >> 2);
+	temp *= cfg->cal2_temp - cfg->cal1_temp;
+	temp /= (*cfg->cal2_addr >> 2) - (*cfg->cal1_addr >> 2);
+	temp += cfg->cal_offset;
+#else
 	temp = ((float)data->raw * cfg->tsv_mv) / cfg->cal_vrefanalog;
 	temp -= *cfg->cal1_addr;
 	temp *= (cfg->cal2_temp - cfg->cal1_temp);
 	temp /= (*cfg->cal2_addr - *cfg->cal1_addr);
 	temp += cfg->cal_offset;
+#endif
```
Can we abstract this difference using device tree ?

FRASTM replied:
Using the soc DTS  should introduce a new property <ADC resolution for TS_CAL> for each ADC with <st_stm32_temp_cal>.
The data TS_CAL1 (*cfg->cal1_addr) and TS_CAL2 (*cfg->cal2_addr) are stored in the system memory of the device. Depending on the SOC, the ADC is 12 or 14 or 16 bits wide and the TS_CAL1 and TS_CAL2 are measured at this max resolution. 
On the other hand the stm32_temp.c is always converting with 12bit resolution :
```
*asp = (struct adc_sequence){
		.resolution = 12U
}
```
That's why the shift is required.

FRASTM replied:
Could also be done with a "custom flag" (like in  the  include/zephyr/dt-bindings/pwm/stm32_pwm.h with PWM_POLARITY_NORMAL  / PWM_STM32_COMPLEMENTARY)
--> TS_CAL_STM32_16BIT / TS_CAL_STM32_14BIT / TS_CAL_STM32_12BIT

erwango replied:
> Could also be done with a "custom flag" (like in the include/zephyr/dt-bindings/pwm/stm32_pwm.h with PWM_POLARITY_NORMAL / PWM_STM32_COMPLEMENTARY) --> TS_CAL_STM32_16BIT / TS_CAL_STM32_14BIT as follows:
> 
> ```
> 		die_temp: dietemp {
> 			compatible = "st,stm32-temp-cal";
> 			ts-cal1-addr = <0x0BFA0710 | TS_CAL_STM32_16BIT>;
> 			ts-cal2-addr = <0x0BFA0742 | TS_CAL_STM32_16BIT>;
> ```

Using a flag on a 32 bits addr field looks a bit fragile. Can't we add a new `ts-cal-resolution` field ?

FRASTM commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -88,7 +88,20 @@ static int stm32_temp_channel_get(const struct device *dev, enum sensor_channel
-#if defined(CONFIG_SOC_SERIES_STM32U5X)
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	/*
```
Please check that in details for stm32h7 devices: stm32h723,  stm32h735, stm32h7b3 etc. There are many ADC version in this soc family.

erwango commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -372,11 +372,23 @@
 		};
 
+		die_temp: dietemp {
+			label = "DIE_TEMP";
```
Nitpick: Please put "compatible" property first. "label" is a property that is close to be obsolete and is nearly not used. As such it feels strange seeing it first.

TDhaouST committed:
dts: binding: sensor: stm32-temp add temp sensor calibration property

The TS_CAL values for the temperature sensor are measured on 12, 14 and
16 bit resolution depends on the STM32 series. Because the drivers
operates at 12 bit resolution the TS_CAL1 and TS_CAL2 must be divided.

Signed-off-by: TLIG Dhaou <dhaou.tlig-ext@st.com>

TDhaouST committed:
drivers: sensor: stm32_temp add the ts cal resolution to the driver

Add the ts cal resolution in the driver in order to calculate temperature
value.

Signed-off-by: TLIG Dhaou <dhaou.tlig-ext@st.com>

pdietl commented:
Nice. I made a PR making _one_ of these changes.

erwango commented on dts/bindings/sensor/st,stm32-temp-cal.yaml:
```diff
@@ -43,3 +43,15 @@ properties:
+      type: int
+      required: false
+      description: |
+        temperature calibration resolution with which the ts-cal1-temp and
+        ts-cal2-temp are measured on 12, 14 and 16 bits.
```
```suggestion
        Temperature calibration resolution with which the ts-cal1-temp and
        ts-cal2-temp are measured.
```

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -143,7 +144,8 @@ static const struct stm32_temp_config stm32_temp_dev_config = {
 	.cal_vrefanalog = DT_INST_PROP(0, ts_cal_vrefanalog),
-	.cal_offset = DT_INST_PROP(0, ts_cal_offset)
+	.cal_offset = DT_INST_PROP(0, ts_cal_offset),
+	.ts_cal_resolution = (DT_INST_PROP(0, ts_cal_resolution) - 12)
```
After doing the substraction `.ts_cal_resolution` should then be called `ts_cal_shift` as it's not a resolution anymore

```suggestion
	.ts_cal_shift = (DT_INST_PROP(0, ts_cal_resolution) - CAL_RES)
```
with
```
#define CAL_RES 12
```
(Prefer use of defines with explicit names rather than "magic"  numbers).
@ABOSTM, @FRASTM do you have an idea for a better name than `CAL_RES`

FRASTM replied:
RAW_DATA_RESOLUTION ?

erwango reviewed (changes requested):
Almost good! Some few more changes required.

erwango commented on dts/bindings/sensor/st,stm32-temp-cal.yaml:
```diff
@@ -43,3 +43,15 @@ properties:
+      description: |
+        Temperature calibration resolution with which the ts-cal1-temp and
+        ts-cal2-temp are measured.
+      default: 12
```
@MaureenHelm, are you good with this justification?

MaureenHelm replied:
Yes

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -853,11 +865,7 @@ static void adc_stm32_setup_speed(const struct device *dev, uint8_t id,
 #if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
 	LL_ADC_SetSamplingTimeCommonChannels(adc,
 		table_samp_time[acq_time_index]);
-#elif defined(CONFIG_SOC_SERIES_STM32G0X)
-	/* Errata ES0418 and more: ADC sampling time might be one cycle longer */
-	if (acq_time_index  < 2) {
-		acq_time_index = 2;
-	}
```
You should not remove this. It was just added recently. See https://github.com/zephyrproject-rtos/zephyr/pull/46125

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -93,14 +93,16 @@ static const uint32_t table_resolution[] = {
+#elif defined(CONFIG_SOC_SERIES_STM32U5X)
+	RES(8),
+	RES(10),
+	RES(12),
+	RES(14),
 #elif !defined(CONFIG_SOC_SERIES_STM32H7X)
 	RES(6),
 	RES(8),
 	RES(10),
 	RES(12),
-#if defined(CONFIG_SOC_SERIES_STM32U5X)
-	RES(14),
-#endif /* CONFIG_SOC_SERIES_STM32U5X */
 #else
```
In a next PR we can use ts-cal-resolution to simplify this table.
Same with it's usage in `start_read()`

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -607,9 +607,21 @@ static int start_read(const struct device *dev,
 	defined(CONFIG_SOC_SERIES_STM32WLX)
 	/* STM32G0 in "not fully configurable" sequencer mode */
 	LL_ADC_REG_SetSequencerChannels(adc, channel);
+	LL_ADC_REG_SetSequencerConfigurable(adc, LL_ADC_REG_SEQ_CONFIGURABLE);
```
this is going to change for stm32G0 too, but this device is preferably configured with 
LL_ADC_REG_SetSequencerConfigurable(adc, **LL_ADC_REG_SEQ_FIXED**);

FRASTM commented on dts/arm/st/g0/stm32g0.dtsi:
```diff
@@ -360,13 +360,26 @@
+			ts-cal2-addr = <0x1FFF75CA>;
+			ts-cal1-temp = <30>;
+			ts-cal2-temp = <130>;
+			ts-cal-vrefanalog = <3300>;
```
please check the Datasheet: 
"TS_CAL1 and TS_CAL2 are raw data acquired at a temperature of 30 °C  and 130 °C   (± 5 °C),  VDDA = VREF+ = **3.0 V** (± 10 mV)"

FRASTM reviewed (commented):
running sample on the stm32g071rb nucleo gives : 
`Current temperature: 29.3 °C     `

TDhaouST commented on drivers/adc/adc_stm32.c:
```diff
@@ -583,7 +593,13 @@ static int start_read(const struct device *dev,
 	 */
 	LL_ADC_SetChannelPreSelection(adc, channel);
 #endif
-
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+	/*
+	 * Each channel in the sequence must be previously enabled in PCSEL.
+	 * This register controls the analog switch integrated in the IO level.
+	 */
+	LL_ADC_SetChannelPreselection(adc, channel);
+#endif
```
No it's not possible to factorize, i realized that it's not the same function LL_ADC_SetChannelPreselection for U5(Preselction) and LL_ADC_SetChannelPreSelection for H7 series (PreSelection).

erwango replied:
Too bad, we need to raise an issue in Cube to get this fixed

MaureenHelm commented:
@ABOSTM please revisit

ABOSTM commented on dts/arm/st/wb/stm32wb.dtsi:
```diff
@@ -357,12 +357,25 @@
 
+		die_temp: dietemp {
+			compatible = "st,stm32-temp-cal";
+			ts-cal1-addr = <0x1FFF75AA>;
```
From STM32WB Datasheet, DS11929 Rev 12, address are wrong:
![image](https://user-images.githubusercontent.com/50730894/178757640-de06e055-cda3-481b-8f36-7b63ce04acf7.png)

I didn't check all Datasheet, but please check all addresses.

ABOSTM commented on dts/arm/st/l5/stm32l5.dtsi:
```diff
@@ -584,12 +584,25 @@
+			ts-cal1-addr = <0x0BFA05A8>;
+			ts-cal2-addr = <0x0BFA05CA>;
+			ts-cal1-temp = <30>;
+			ts-cal2-temp = <110>;
```
From datasheet, DS12737 Rev 6, Cal2 temp is 130°c not 110:
![image](https://user-images.githubusercontent.com/50730894/178760314-0879c3d5-5acd-49b8-95ff-0d522e69cf81.png)

ABOSTM commented on dts/arm/st/wb/stm32wb.dtsi:
```diff
@@ -357,12 +357,25 @@
+			ts-cal1-addr = <0x1FFF75AA>;
+			ts-cal2-addr = <0x1FFF75AB>;
+			ts-cal1-temp = <30>;
+			ts-cal2-temp = <110>;
```
`ts-cal2-temp` should be 130°c
Please check them all

ABOSTM commented on dts/arm/st/f7/stm32f7.dtsi:
```diff
@@ -688,12 +688,23 @@
 		};
 
+		die_temp: dietemp {
+			compatible = "st,stm32-temp";
```
dito

ABOSTM commented on dts/arm/st/l4/stm32l4.dtsi:
```diff
@@ -371,10 +371,23 @@
+			ts-cal1-addr = <0x1FFF75A8>;
+			ts-cal2-addr = <0x1FFF75CA>;
+			ts-cal1-temp = <30>;
+			ts-cal2-temp = <130>;
```
```suggestion
			ts-cal2-temp = <110>;
```
`Please check them all`

ABOSTM commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -676,6 +676,20 @@
+			ts-cal2-temp = <110>;
+			ts-cal-vrefanalog = <3300>;
+			ts-cal-resolution = <16>;
+			io-channels = <&adc3 18>;
```
Unfortunately, channel is not the same for H7 series:
for example H723/H725/H730 is channel 17
Please check channel for all H7 MCU.

ABOSTM commented on dts/arm/st/h7/stm32h723.dtsi:
```diff
@@ -30,6 +30,10 @@
 		};
 
+		die_temp: dietemp {
+			io-channels = <&adc3 17>;
```
H723 has TS_CAL2 130°c

ABOSTM commented on dts/arm/st/h7/stm32h7a3.dtsi:
```diff
@@ -58,6 +58,10 @@
 		};
 
+		die_temp: dietemp {
+			io-channels = <&adc2 18>;
```
H7A3 has TS_cal2 130°c
also address are different

ABOSTM reviewed (dismissed):
LGTM

FRASTM reviewed (dismissed):
LGTM 
but should this PR about ADC temp sensor mix commits 1e614d2c1701a41e462190a177b305d206e696b0 and eaf09a885bb07f30501d48f5c241dad84e4e723d about _drivers: adc: stm32 init the Regular group_

carlescufi commented:
@TDhaouST please rebase
